### PR TITLE
Feature tmc#144 floats in channels

### PR DESCRIPTION
**Pull request by gawkowsk**
a lot of todo left for further work, but it seems to work on quite sophisticated scenarios...

### CODE REVIEW

File: Test/Float/ComplexFloatProcessing.pml [hunk 1]
```diff
@@ -39,36 +47,69 @@ inline random10(x) {
...
-    ::  random10(msg.index);
-        randomFloat(msg.value);
+    ::count < 10 ->
+        
+        //random10(msg.index);
+        //randomFloat(msg.value);
+        msg.index = count;
+        msg.value = count/10.0 + 0.1;
+        if
+        :: msg.value == 1.0 -> msg.value = 0.75;
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml]: Why 1.0 := 0.75?

File: Test/Float/ComplexFloatProcessing.pml [hunk 2]
```diff
@@ -39,36 +47,69 @@ inline random10(x) {
...
+        fi
+        printf("Sender messsage: index=%d, value=%f\n", msg.index, msg.value);
         queue!msg;
+        count++;
+    ::else -> break;
     od
+
+    msg.index = finalIndex;
+    msg.value = finalValue;
+    queue!msg;
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml]: Why 3 times?

File: Test/Float/ComplexFloatProcessing.pml [hunk 3]
```diff
@@ -39,36 +47,69 @@ inline random10(x) {
...
 
 active proctype sender() {
     Message msg;
+    byte count;
     do
-    ::  random10(msg.index);
-        randomFloat(msg.value);
+    ::count < 10 ->
+        
+        //random10(msg.index);
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml]: Please remove commented out code.

File: Test/Float/ComplexFloatProcessing.pml [hunk 4]
```diff
@@ -39,36 +47,69 @@ inline random10(x) {
 
 inline randomFloat(x) {
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml:27]: This is unused - please remove.

File: Test/Float/ComplexFloatProcessing.pml [hunk 5]
```diff
@@ -17,10 +8,27 @@ typedef Message {
...
+    printf("data[3]=%f\n", data.values[3]);
+    printf("data[4]=%f\n", data.values[4]);
+    printf("data[5]=%f\n", data.values[5]);
+    printf("data[6]=%f\n", data.values[6]);
+    printf("data[7]=%f\n", data.values[7]);
+    printf("data[8]=%f\n", data.values[8]);
+    printf("data[9]=%f\n", data.values[9]);
+}
 
 inline random10(x) {
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml:12]: This is unused - please remove.

File: Test/Float/ComplexFloatProcessing.pml [hunk 6]
```diff
@@ -17,10 +8,27 @@ typedef Message {
 
 typedef Data {
     float values[STORE_SIZE];
+    int intValue;
```
Comment (Lurkerpas) [inline Test/Float/ComplexFloatProcessing.pml]: Why is this added?

File: Test/Float/Channels/ChannelsWithFloatsCanBeEnumerated.pml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    q!3.14;
+    float x = 2.7;
+    q!x;
+    synchro!1;
+}
+
+active proctype receiver()
+{
+    m a;
+    //q?a;
```
Comment (Lurkerpas) [inline Test/Float/Channels/ChannelsWithFloatsCanBeEnumerated.pml]: Please remove commented out code.

File: Test/Float/Channels/ComplexFloatMessage.pml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    ::queue?msg;
+        printf("Receiver: index=%d, value=%f\n", msg.index, msg.value);
+        if
+        ::(msg.index == finalIndex && msg.value == finalValue) -> 
+            printf("Final message received!\n");
+            break;
+        ::else -> 
+            if 
+            ::msg.index < STORE_SIZE -> 
+                data[0].values[msg.index] = msg.value;
```
Comment (Lurkerpas) [inline Test/Float/Channels/ComplexFloatMessage.pml:37]: I do not understand the intent of this test.  Please either refactor it, remove it, or provide comments.

File: Test/Float/Channels/ComplexFloatMessage.pml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                data[1].floatValue = data[1].floatValue  + msg.value; 
+                data[2].intValue = data[2].intValue + msg.index;
+            ::else ->
+                printf("Erronous index = %d\n", msg.index);
+                break;
+            fi
+        fi
+        
+    od
+    dqueue!data[0];
```
Comment (Lurkerpas) [inline Test/Float/Channels/ComplexFloatMessage.pml:48]: Why?

File: Test/Float/Channels/ComplexFloatMessage.pml (hunk 8 above)
Comment (Lurkerpas) [reply Test/Float/Channels/ComplexFloatMessage.pml:37]: If this is to test a complex message, then I believe that the test should be more explicit on that and focus on the complex message itself.

File: Test/Float/Dumping/FloatsCanBeDumped.pml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    destination.inner_int = origin.inner_int;
+
+    copying_channel!origin.floats;
+    copying_channel?destination.floats;
+}
+
+active proctype test_dump(){
+    float_var = 4.83;
+    byte_var = 6;
+    int_var = 234565;
```
Comment (Lurkerpas) [inline Test/Float/Dumping/FloatsCanBeDumped.pml:62]: I think this test could use some comments. Also, I think it might test a few separate things. Please consider splitting into two or three.

File: Src/sym.c [hunk 11]
```diff
@@ -159,11 +159,21 @@ trackvar(Lextok *n, Lextok *m)
 				sp->hidden |= 8;
 		}
 		break;
-	case CONST:		// TODO PG - varify for floats
-		if (m->val != 0 && m->val != 1)
-			sp->hidden |= 4;
-		if (m->val < 0 || m->val > 256)
-			sp->hidden |= 8; /* ditto byte-equiv */
+	case CONST:		// TODO PG - not sure if float const are well handled here
```
Comment (Lurkerpas) [inline Src/sym.c:162]: It is my suspicion that this has no sense for floats, as 256.0 is not a special number.... But I am not sure what this code does either. I do not expect any changes, but it would be good to know what it does.

File: Src/pangen3.c [hunk 12]
```diff
@@ -28,8 +28,7 @@ static SRC	*frst = (SRC *) 0;
 static SRC	*skip = (SRC *) 0;
 
 extern int	ltl_mode;
-extern void sr_mesg_f(FILE *, float, int, const char *);
-extern void	sr_mesg(FILE *, int, int, const char *);
+extern void sr_mesg_const_token(FILE *,const Lextok *, const char *);
```
Comment (Lurkerpas) [inline Src/pangen3.c]: Ident

File: Src/mesg.c [hunk 13]
```diff
@@ -38,8 +38,8 @@ static int	sa_snd(Queue *, Lextok *);
 static int	s_snd(Queue *, Lextok *);
 extern Lextok	**find_mtype_list(const char *);
 extern char	*which_mtype(const char *);
-extern void	sr_buf(int, int, const char *);
-extern void	sr_mesg(FILE *, int, int, const char *);
+void	sr_buf(int, int, const char *);
```
Comment (Lurkerpas) [inline Src/mesg.c:41]: Indent.

File: Test/Makefile [hunk 14]
```diff
@@ -511,3 +517,95 @@ test_float_in_struct:
...
+	grep -q "154 states, matched" ${TEST_DIR}/ComplexFloatMessage.output
+	grep -q "hash conflicts:         0 (resolved)" ${TEST_DIR}/ComplexFloatMessage.output
+	grep -q "Receiver: index=11, value=0.250000" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "Final message received!" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "queue 1 (queue): \[11,0.250000\]\[11,0.300000\]" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "queue 2 (dqueue): \[0.100000,0.200000,0.300000,0.400000,0.500000,0.600000,0.700000,0.800000,0.900000,0.750000,0,0.000000,52\]\[0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0,5.250000,0\]\[0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,45,0.000000,0\]" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "finalValue = 0.250000" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "finalIndex = 11" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "ComplexFloatMessage.pml:80 (state 25)" ${TEST_DIR}/ComplexFloatMessage.sim.output
+	grep -q "ComplexFloatMessage.pml:51 (state 25) <valid end state>" ${TEST_DIR}/ComplexFloatMessage.sim.output
```
Comment (Lurkerpas) [inline Test/Makefile:621]: I don't think we need to test the end states.

File: Test/Float/Channels/SimpleFloatThroughChannel.pml [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    /* float literals can be sent explicitly */
+    q!3.14;
+    /* floats can be sent as variable */
+    float x = 2.7;
+    q!x;
+}
+
+active proctype receiver()
+{
+    float a;
```
Comment (Lurkerpas) [inline Test/Float/Channels/SimpleFloatThroughChannel.pml:19]: This is a general comment for the new tests, except the "Dumping" one. The "sim" output verifies the prints. However, the model checker output is verified only for errors, but there are no (or no meaningful) checks. It would be great if:

- there was a positive test, which does not yield an error, despite incorporating assertions on the received data.
- there was a negative test, which yields an error on an intentional unsatisfied assert.

```
E.g., :
active proctype bad_receiver(){
float a;
q?a;
printf("first value = %f\n", a);
assert(a > 3.5);
}


```

File: Test/Float/ComplexFloatProcessing.pml (hunk 1 above)
Comment (gawkowsk) [reply Test/Float/ComplexFloatProcessing.pml]: nevermind - ComplexFloatProcessing is reverted now...

File: Test/Float/ComplexFloatProcessing.pml (hunk 2 above)
Comment (gawkowsk) [reply Test/Float/ComplexFloatProcessing.pml]: nevermind - ComplexFloatProcessing is reverted now...

File: Test/Float/ComplexFloatProcessing.pml (hunk 3 above)
Comment (gawkowsk) [reply Test/Float/ComplexFloatProcessing.pml]: nevermind - ComplexFloatProcessing is reverted now...

File: Test/Float/ComplexFloatProcessing.pml (hunk 5 above)
Comment (gawkowsk) [reply Test/Float/ComplexFloatProcessing.pml:12]: nevermind - ComplexFloatProcessing is reverted now...

File: Test/Float/ComplexFloatProcessing.pml (hunk 6 above)
Comment (gawkowsk) [reply Test/Float/ComplexFloatProcessing.pml]: nevermind - ComplexFloatProcessing is reverted now...

File: Test/Float/Channels/ChannelsWithFloatsCanBeEnumerated.pml (hunk 7 above)
Comment (gawkowsk) [reply Test/Float/Channels/ChannelsWithFloatsCanBeEnumerated.pml]: done

File: Test/Makefile (hunk 14 above)
Comment (gawkowsk) [reply Test/Makefile:621]: in this particular test, the data was intentionally dumped to get all global vars. I've used blocked message sending for that purpose, so, the receiver should go to the valid end state while the sender hangs-up. That is why I verify end states here

File: Src/mesg.c (hunk 13 above)
Comment (gawkowsk) [reply Src/mesg.c:41]: in fact, it is indented the way it should be - the convention is that the fun name is separated with tab from its type - here, there is no need for "extern", so, it looks the same as, for instance, decl. in line 29 or 30

File: Src/sym.c (hunk 11 above)
Comment (gawkowsk) [reply Src/sym.c:162]: unfortunately, the "hidden" field consists of several flags (with different meanings). one of them says "hey, I have hidden attribute". But some of them says "hey, you can cast my value to bit or byte" - that is why I changed that code to handle with floats...

File: Test/Float/Channels/ComplexFloatMessage.pml (hunk 8 above)
Comment (gawkowsk) [reply Test/Float/Channels/ComplexFloatMessage.pml:37]: both queues use a different message types while the test itself use quite complex logic to send several messages and then the final message which end the communication. sending messages of the same type with different data tests if the channel really sends the right data in subsequent actions.

Comment (gawkowsk) [reply Test/Float/Channels/ComplexFloatMessage.pml:37]: putting different data (calculating a simple sum from different fields of the message) into separate array elements verifies if the array elements are really separated from each other....

File: Test/Float/Dumping/FloatsCanBeDumped.pml (hunk 10 above)
Comment (gawkowsk) [reply Test/Float/Dumping/FloatsCanBeDumped.pml:62]: Yep - using channel for copying complex structures could be in another test...  lack of time... sorry

File: Test/Float/Channels/ComplexFloatMessage.pml (hunk 8 above)
Comment (Lurkerpas) [reply Test/Float/Channels/ComplexFloatMessage.pml:37]: Still, the code is not clear. If it is to be kept, then it should be either refactored, or at least commented with respect to the intent. The intended results of this test are not clear for my perspective.

File: Src/mesg.c [hunk 16]
```diff
@@ -489,11 +491,11 @@ s_snd(Queue *q, Lextok *n)
...
 
 		for (j = 0; m && j < q->nflds; m = m->rgt, j++)
-		{	sr_talk(n, eval(m->lft), "Sent ", "->", j, q);
+		{	sr_talk(n, evalValue(m->lft), "Sent ", "->", j, q);
 		}
-
-		for (i = j; i < q->nflds; i++)
-		{	sr_talk(n, 0, "Sent ", "->", i, q);
+		int tmp;
+		for (tmp = j; tmp < q->nflds; tmp++)
```
Comment (Lurkerpas) [inline Src/mesg.c:497]: Why change i to tmp?

File: Test/Float/Channels/SimpleFloatThroughChannel.pml [hunk 17]
```diff
@@ -1,19 +1,30 @@
...
 chan q = [5] of {float};
 
 active proctype sender()
 {
-    /* float literals can be sent explicitly */
-    q!3.14;
-    /* floats can be sent as variable */
-    float x = 2.7;
+    
+    /* int literals can be sent explicitly */
```
Comment (Lurkerpas) [inline Test/Float/Channels/SimpleFloatThroughChannel.pml:9]: Why change float to int in the comment?

File: Test/Makefile [hunk 18]
```diff
@@ -563,14 +564,25 @@ test_simple_floats_through_channel:
...
+	grep -q "second value = 3.140000" ${TEST_DIR}/SimpleFloatThroughChannel.sim.output
 	grep -q "2 processes created" ${TEST_DIR}/SimpleFloatThroughChannel.sim.output
 	${DELETE_TEST_DIR}
 
+#Test ping pong with floats through channel 
+test_ping_pong_with_floats_through_channel:
+	echo "Test ping pong with floats through channel"
+	${CREATE_TEST_DIR}
+	cd ${TEST_DIR} && \
+	${SPIN} -run ${FLOAT_CHANNEL_MODEL_DIR}/PingPong.pml > PingPong.output
```
Comment (Lurkerpas) [inline Test/Makefile:583]: Can we test the output of this verification?

Comment (Lurkerpas) [reply Test/Makefile:583]: Rejected as "won't fix"

File: Test/Float/Dumping/FloatsCanBeDumped.pml (hunk 10 above)
Comment (Lurkerpas) [reply Test/Float/Dumping/FloatsCanBeDumped.pml:62]: Rejected as "won't fix"

File: Test/Float/Channels/SimpleFloatThroughChannel.pml (hunk 15 above)
Comment (Lurkerpas) [reply Test/Float/Channels/SimpleFloatThroughChannel.pml:19]: Fixed for some cases

File: Test/Float/Channels/SimpleFloatThroughChannel.pml (hunk 17 above)
Comment (Lurkerpas) [reply Test/Float/Channels/SimpleFloatThroughChannel.pml:9]: Rejected as "won't fix"

File: Test/Float/Channels/ComplexFloatMessage.pml (hunk 9 above)
Comment (Lurkerpas) [reply Test/Float/Channels/ComplexFloatMessage.pml:48]: No answer

File: Test/Float/Channels/ComplexFloatMessage.pml (hunk 8 above)
Comment (Lurkerpas) [reply Test/Float/Channels/ComplexFloatMessage.pml:37]: Rejected as "won't fix"

File: Src/mesg.c (hunk 16 above)
Comment (Lurkerpas) [reply Src/mesg.c:497]: No asnwer

**Review by Lurkerpas — APPROVED** [review]
